### PR TITLE
Cache fixes for high concurrency on corner cases

### PR DESCRIPTION
**Pull request by kssenii**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Cache fixes for high concurrency on corner cases.

### CODE REVIEW

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 1]
```diff
@@ -63,7 +63,11 @@ void CachedReadBufferFromRemoteFS::initialize(size_t offset, size_t size)
 
 SeekableReadBufferPtr CachedReadBufferFromRemoteFS::getCacheReadBuffer(size_t offset) const
 {
-    return std::make_shared<ReadBufferFromFile>(cache->getPathInLocalCache(cache_key, offset), settings.local_fs_buffer_size);
+    auto path = cache->getPathInLocalCache(cache_key, offset);
+    auto buf = std::make_shared<ReadBufferFromFile>(path, settings.local_fs_buffer_size);
+    if (buf->size() == 0)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Attempt to read from an empty cache file: {}", path);
```
Comment (kssenii) [inline src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:69]: better to put these 3 lines this under NDEBUG?

File: src/Common/FileSegment.cpp [hunk 2]
```diff
@@ -439,6 +495,56 @@ String FileSegment::stateToString(FileSegment::State state)
...
+{
+    /// In CacheableReadBufferFromRemoteFS file segment's downloader removes file segments from
+    /// FileSegmentsHolder right after calling file_segment->complete(), so on destruction here
+    /// remain only uncompleted file segments.
+
+    IFileCache * cache = nullptr;
+
+    for (auto file_segment_it = file_segments.begin(); file_segment_it != file_segments.end();)
+    {
+        auto current_file_segment_it = file_segment_it++;
```
Comment (alesapin) [inline src/Common/FileSegment.cpp:521]: Check comment bellow.

File: src/Common/FileSegment.cpp [hunk 3]
```diff
@@ -439,6 +495,56 @@ String FileSegment::stateToString(FileSegment::State state)
...
+
+        try
+        {
+            /// File segment pointer must be reset right after calling complete() and
+            /// under the same mutex, because complete() checks for segment pointers.
+            std::lock_guard cache_lock(cache->mutex);
+
+            file_segment->complete(cache_lock);
+
+            file_segments.erase(current_file_segment_it);
```
Comment (alesapin) [inline src/Common/FileSegment.cpp:535]: better to make:
`file_segment_it = file_segments.erase(file_segment_it);`

File: src/Common/FileSegment.cpp [hunk 4]
```diff
@@ -299,107 +332,130 @@ void FileSegment::completeBatchAndResetDownloader()
...
-void FileSegment::completeImpl(bool allow_non_strict_checking)
+void FileSegment::completeImpl(std::lock_guard<std::mutex> & cache_lock, std::lock_guard<std::mutex> & segment_lock, bool allow_non_strict_checking)
 {
-    /// cache lock is always taken before segment lock.
-    std::lock_guard cache_lock(cache->mutex);
-    std::lock_guard segment_lock(mutex);
-
-    bool download_can_continue = false;
+    bool is_last_holder = cache->isLastFileSegmentHolder(key(), offset(), cache_lock, segment_lock);
+    bool download_can_continue = !is_last_holder
```
Comment (alesapin) [inline src/Common/FileSegment.cpp:414]: Remove

File: src/Common/FileSegment.cpp [hunk 5]
```diff
@@ -299,107 +332,130 @@ void FileSegment::completeBatchAndResetDownloader()
...
+        && download_state != State::DOWNLOADED
+        && download_state != State::PARTIALLY_DOWNLOADED_NO_CONTINUATION;
 
-    if (download_state == State::PARTIALLY_DOWNLOADED
-                || download_state == State::PARTIALLY_DOWNLOADED_NO_CONTINUATION)
+    if (!download_can_continue
```
Comment (alesapin) [inline src/Common/FileSegment.cpp:418]:
```c++
if (download_state == State::PARTIALLY_DOWNLOADED || download_state == State::PARTIALLY_DOWNLOADED_NO_CONTINUATION)
{
    if (current_downloaded_size == 0)
    {
          /// TODO
    }
    else if (is_last_holder)
    {
    }

    if (cache_writer)
}
```

**Review by alesapin — APPROVED** [review]

File: src/Disks/IO/ThreadPoolRemoteFSReader.cpp [hunk 6]
```diff
@@ -1,5 +1,6 @@
 #include "ThreadPoolRemoteFSReader.h"
 
+#include <base/scope_guard_safe.h>
```
Comment (alesapin) [inline src/Disks/IO/ThreadPoolRemoteFSReader.cpp:3]: Redundant?

**Comment by alesapin** [comment]
@Mergifyio update

**Comment by kssenii** [comment]
> Integration tests (asan, actions) [2/3] — fail: 2

-- test_mysql_protocol/test.py::test_php_client - not related
-- test_odbc_interaction/test.py::test_bridge_dies_with_parent - not related

> ClickHouse special build check (actions) — 7/7 builds are OK

:monocle_face:

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 7]
```diff
@@ -618,6 +634,18 @@ bool CachedReadBufferFromRemoteFS::nextImplStep()
 
     if (!result)
     {
+#ifdef NDEBUG
```
Comment (alesapin) [inline src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:637]:
```suggestion
#ifndef NDEBUG
```

File: src/Disks/IO/CachedReadBufferFromRemoteFS.cpp [hunk 8]
```diff
@@ -296,17 +298,21 @@ SeekableReadBufferPtr CachedReadBufferFromRemoteFS::getImplementationBuffer(File
     {
         case ReadType::CACHED:
         {
-            size_t seek_offset = file_offset_of_buffer_end - range.left;
-            read_buffer_for_file_segment->seek(seek_offset, SEEK_SET);
-
+#ifdef NDEBUG
```
Comment (alesapin) [inline src/Disks/IO/CachedReadBufferFromRemoteFS.cpp:301]:
```suggestion
#ifndef NDEBUG
```

**Comment by alesapin** [comment]
No related failures:
1. In func tests `AWSClient: Failed to make request to: http://localhost:11111/test/test/fxqhfbyqaacqjblwdqjpmuslwbktgrdb`. Something wrong with minio.
2. Integration tests test_mysql_protocol and jemalloc test.

**Pull request merged**